cpp/src/buffer/buffer.cpp

```diff
@@ -18,20 +20,62 @@ template <typename T>
 }
 }  // namespace
 
+Buffer::Event::Event(rmm::cuda_stream_view stream) {
+    RAPIDSMPF_CUDA_TRY(cudaEventCreateWithFlags(&event_, cudaEventDisableTiming));
+    RAPIDSMPF_CUDA_TRY(cudaEventRecord(event_, stream));
+}
+
+Buffer::Event::~Event() {
+    // Mark as destroying - if we fail, another thread is already destroying
+    bool expected = false;
+    if (!destroying_.compare_exchange_strong(expected, true)) {
+        return;
+    }
+
+    // Finally acquire the mutex and destroy the event
+    std::lock_guard<std::mutex> lock(mutex_);
+    cudaEventDestroy(event_);
```

Comment:
I feel like this logic is loose. No one's setting `destroying_` value, right?

Reply:
I think logic becomes much more simpler if you just used a mutex and 2 bools. We dont want to destroy the same event twice.

Reply:
Sorry I stand corrected, CES does store it to true, here.